### PR TITLE
Batch reading, writing and deleting of chunks

### PR DESCRIPTION
**Pull request by bboreham**
This rather large PR is almost everything I did to be able to re-index a year of data on DynamoDB.

Because of the way DynamoDB is priced, the best way to do a batch run is to go at a very steady speed.  Here I have made a lot of use of rate-limiters and stopped using exponential backoff.

I have taken out most of the janky bits and blind alleys, but left in quite a lot of the journey in case it helps understanding.  I may have lost something vital from the `scanner` program in the rebase against master - it came from a long way back - but tests are passing.

I'm not expecting this to be merged as-is: I put it here for visibility and because I think the batched writer is a better way to do chunk writes from the Ingester but I haven't plumbed them together yet.

### CODE REVIEW

File: pkg/chunk/aws/storage_client.go [hunk 1]
```diff
@@ -713,6 +761,68 @@ func (b dynamoDBWriteBatch) Add(tableName, hashValue string, rangeValue []byte,
 	})
 }
 
+// FIXME: temporary hack passing in storageClient so we can get at the schemaCfg
```
Comment (tomwilkie) [inline pkg/chunk/aws/storage_client.go]: FIXME

Also keen not add more dependancies from here to the `chunk` pkg, this dependancy is something I want to remove in the future.

File: cmd/scanner/main.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	checkFatal(err)
+	defer chunkStore.Stop()
+
+	var reindexStore chunk.Store
+	if reindexTablePrefix != "" || rechunkTablePrefix != "" {
+		reindexSchemaConfig := schemaConfig
+		//FIXME
+		//reindexSchemaConfig.ChunkTables.Prefix = rechunkTablePrefix
+		//reindexSchemaConfig.IndexTables.Prefix = reindexTablePrefix
+		reindexStore, err = storage.NewStore(storageConfig, chunkStoreConfig, reindexSchemaConfig, overrides)
```
Comment (tomwilkie) [inline cmd/scanner/main.go]: FIXME?

File: cmd/scanner/main.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+		checkFatal(err)
+	}
+
+	tableName = "FIXME" //fmt.Sprintf("%s%d", schemaConfig.ChunkTables.Prefix, week)
```
Comment (tomwilkie) [inline cmd/scanner/main.go]: FIXME?

File: pkg/chunk/chunk_store.go [hunk 4]
```diff
@@ -118,29 +134,25 @@ func (c *store) Put(ctx context.Context, chunks []Chunk) error {
...
 	}
+	batch := c.storage.NewWriteBatch()
+	batch.AddChunk(c.storage, chunk, buf)
+	c.writer.Write <- batch
 
-	// Horribly, PutChunks mutates the chunk by setting its checksum.  By putting
-	// the chunk in a slice we are in fact passing by reference, so below we
-	// need to make sure we pick the chunk back out the slice.
-	chunks := []Chunk{chunk}
+	// FIXME: shouldn't send chunk to the cache until it has gone to the real store
```
Comment (tomwilkie) [inline pkg/chunk/chunk_store.go]: FIXME

File: pkg/chunk/aws/storage_client.go (hunk 1 above)
Comment (tomwilkie) [reply pkg/chunk/aws/storage_client.go]: Yeah, looks like only used in `PutOne`, so the logic could go there?

File: pkg/chunk/chunk_store.go [hunk 5]
```diff
@@ -118,29 +134,25 @@ func (c *store) Put(ctx context.Context, chunks []Chunk) error {
...
 func (c *store) PutOne(ctx context.Context, from, through model.Time, chunk Chunk) error {
-	userID, err := user.ExtractOrgID(ctx)
+	// Encode the chunk first - checksum is calculated as a side effect.
+	buf, err := chunk.Encode()
 	if err != nil {
 		return err
 	}
+	batch := c.storage.NewWriteBatch()
+	batch.AddChunk(c.storage, chunk, buf)
+	c.writer.Write <- batch
```
Comment (tomwilkie) [inline pkg/chunk/chunk_store.go]: This is assuming that the chunks are put in a table-like store - how would this work with S3 (and the upcoming GCS?)

Comment (tomwilkie) [reply pkg/chunk/chunk_store.go]: Also, having the interface to a Writer be a put to a channel seems unfriendly; perhaps this should be a function?

**Comment by tomwilkie** [comment]
Done a first pass.

Minor: Looks like a bunch of fixmes still need addressing.

Bigger issue though, IIUC this makes the chunk store write path async - and therefore it is possible that a chunk is removed the ingesters data structures before it is flushed to the store?

**Comment by bboreham** [comment]
Closed pending rework

**Pull request closed without merging**